src/precompiles/mod.rs

```diff
@@ -106,216 +107,336 @@ impl HardFork for Istanbul {}
-            input, target_gas, context, state, is_static,
-        )),
-        SHA256::<AuroraStackState>::ADDRESS => Some(SHA256::<AuroraStackState>::run(
-            input, target_gas, context, state, is_static,
-        )),
-        RIPEMD160::<AuroraStackState>::ADDRESS => Some(RIPEMD160::<AuroraStackState>::run(
-            input, target_gas, context, state, is_static,
-        )),
-        ExitToNear::<AuroraStackState>::ADDRESS => Some(ExitToNear::<AuroraStackState>::run(
-            input, target_gas, context, state, is_static,
-        )),
-        ExitToEthereum::<AuroraStackState>::ADDRESS => Some(
-            ExitToEthereum::<AuroraStackState>::run(input, target_gas, context, state, is_static),
-        ),
-        _ => None,
-    };
-    output.map(|res| res.map(Into::into))
+impl<'backend, 'config> executor::Precompiles<AuroraStackState<'backend, 'config>>
+    for PrecompileAddresses<Homestead>
+{
+    fn run(
+        &self,
+        address: H160,
+        input: &[u8],
+        target_gas: Option<u64>,
+        context: &Context,
+        state: &mut AuroraStackState,
+        is_static: bool,
+    ) -> Option<EvmPrecompileResult> {
+        let target_gas = match target_gas {
+            Some(t) => t,
+            None => return Some(EvmPrecompileResult::Err(ExitError::OutOfGas)),
+        };
+
+        let output = match address.0 {
+            ECRecover::<AuroraStackState>::ADDRESS => {
+                Some(ECRecover::run(input, target_gas, context, state, is_static))
+            }
+            SHA256::<AuroraStackState>::ADDRESS => {
+                Some(SHA256::run(input, target_gas, context, state, is_static))
+            }
+            RIPEMD160::<AuroraStackState>::ADDRESS => {
+                Some(RIPEMD160::run(input, target_gas, context, state, is_static))
+            }
+            ExitToNear::<AuroraStackState>::ADDRESS => Some(ExitToNear::run(
+                input, target_gas, context, state, is_static,
+            )),
+            ExitToEthereum::<AuroraStackState>::ADDRESS => {
+                Some(ExitToEthereum::<AuroraStackState>::run(
+                    input, target_gas, context, state, is_static,
+                ))
+            }
+            _ => None,
+        };
+        output.map(|res| res.map(Into::into))
+    }
+
+    fn addresses(&self) -> &[H160] {
+        &self.addresses
+    }
+}
+
+impl PrecompileAddresses<Homestead> {
+    #[allow(dead_code)]
+    pub fn homestead() -> Self {
+        Self {
+            addresses: vec![
+                ECRecover::<AuroraStackState>::ADDRESS,
+                SHA256::<AuroraStackState>::ADDRESS,
+                RIPEMD160::<AuroraStackState>::ADDRESS,
+                ExitToNear::<AuroraStackState>::ADDRESS,
+                ExitToEthereum::<AuroraStackState>::ADDRESS,
+            ]
+            .into_iter()
+            .map(Address)
+            .collect(),
```

Comment:
I don't think there will be any impact on gas because

(a) the ones which are never used will be optimized out of the wasm artifact on compilation (since we minimize code size), and

(b) making Istanbul addresses is not necessary, but I don't think it will ever be done in production because this code will likely only be deployed after we actually switch to Berlin.

I included the addresses in the other hard forks for conceptual completeness. That said, if you don't think that conceptual completeness is adding value to people looking at our codebase then I'm fine removing them.

Reply:
It gives me an idea of how we possibly should be doing these. Ala, builders with different EIPs. That would be a fantastic improvement and allow us to be more adaptable with other projects i.e BSC if they choose to have their own improvements to their protocol, or not include some of Ethereum's.

I am working on a bit of a change to this, I'm about 95% happy with it, but an obvious choice for me would be to:
a) Have precompiles in it's own lib
b) Have builders which allows us to choose our own fork or individual EIPs

That would in the end save us time if we choose to connect to ETH forks or other EVMs.